### PR TITLE
The garbage collector doesn't waste time anymore

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Changes the garbage collectors for loop from a `for(var/thing in list)` to `for(var/i in 1 to length(list))`
Converts the garbage queues from an assoc list of id to time, to a list of lists containing the same info.

## Why It's Good For The Game

Did you know that `for(var/thing in list)` scales in cost with the size of the list?
Did you know that assoc accesses are slower then indexed accesses scaling with the size of the list?
Did you know that the garbage collector queue balloons to massive size quite easily?

Note, this means harddels will happen more often, but that's only because the gc list will actually fully process more often.

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
Pending MSO

**Review by MrsTonedOne — APPROVED** [review]

File: code/controllers/subsystem/garbage.dm [hunk 1]
```diff
@@ -139,7 +139,8 @@ SUBSYSTEM_DEF(garbage)
 
 	lastlevel = level
 
-	for (var/refID in queue)
+	for (var/i in 1 to length(queue))
```
Comment (Jared-Fogle) [inline code/controllers/subsystem/garbage.dm:142]: Can you leave a comment explaining why this isnt for...in?

Comment (LemonInTheDark) [reply code/controllers/subsystem/garbage.dm:142]: for you, anything

**Review by Mothblocks — DISMISSED** [review]

**Review by MrsTonedOne — CHANGES_REQUESTED** [review]
https://github.com/LemonInTheDark/tgstation/pull/3

To deal with the fact that multiple places remove items from the list, this pr removes the one other place that was removing from the list

**Comment by LemonInTheDark** [comment]
MSO webeditted, but if it makes you feel better I tested the webedit

**Review by MrsTonedOne — APPROVED** [review]

**Comment by hex37** [comment]
I code for a different SS13 server, but a friend handed me this PR to look at today. Honest to god never knew Dreammaker treated their for loops that poorly.

I gotta go redo all my list traversals now, fuck you, also thank you.

**Comment by LemonInTheDark** [comment]
yello bestie
it's worth noting that it's only really important for really large lists, most of the time it barely matters.
Oh also uh, it ends up being slightly cheaper then for(var/i = 0; i <= list.len; i++) and i think maybe for(var/i in 1 to list.len) too. 
Don't need to do any mass conversions, it mostly matters for lists you don't actually fully transverse, which the gc list is like, defined by

**Pull request merged**